server/filestore.go

```diff
@@ -8743,6 +8772,10 @@ func (fs *fileStore) _writeFullState(force bool) error {
 	// Protect with dios.
 	<-dios
 	err := os.WriteFile(fn, buf, defaultFilePerms)
+	// if file system is not writable isPermissionError is set to true
+	if isPermissionError(err) {
```

Comment:
This **must** be done after the `dios <- struct{}{}` line, otherwise the semaphore becomes unbalanced in a failure scenario.